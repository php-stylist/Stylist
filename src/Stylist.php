<?php declare(strict_types = 1);

namespace Stylist;

use Nette\Utils\Finder;
use Stylist\Checks\CheckInterface;
use Stylist\Output\OutputInterface;


final class Stylist
{

	/** @var CheckInterface[] */
	private $checks;

	/** @var OutputInterface */
	private $output;

	/** @var FileFactory */
	private $fileFactory;

	/** @var string[] */
	private $accepted = [];

	/** @var string[] */
	private $excluded = [];

	/** @var File[] */
	private $files = [];


	/**
	 * @param CheckInterface[] $checks
	 */
	public function __construct(
		array $checks,
		OutputInterface $output,
		FileFactory $fileFactory
	)
	{
		$this->checks = $checks;
		$this->output = $output;
		$this->fileFactory = $fileFactory;
	}


	public function accept(array $accepted): self
	{
		$this->accepted = $accepted;
		return $this;
	}


	public function exclude(array $excluded): self
	{
		$this->excluded = $excluded;
		return $this;
	}


	/**
	 * @param string[] $paths
	 */
	public function check(array $paths): CheckResult
	{
		\set_time_limit(0);

		$this->output->initialize($paths);

		$start = \microtime(true);
		$success = true;
		$checkedFileNames = [];

		foreach ($this->getFiles($paths) as $file) {
			if (\in_array($file->getPathname(), $checkedFileNames, true)) {
				continue;
			}

			$checkedFileNames[] = $file->getPathname();
			$this->files[] = $checkedFile = $this->fileFactory->create($file);
			$success = $this->checkFile($checkedFile) && $success;
			$this->output->checkedFile($checkedFile);
		}

		$timeTaken = \microtime(true) - $start;
		$result = new CheckResult(
			$success,
			$this->files,
			$timeTaken
		);

		$this->output->finish($result);
		return $result;
	}


	/**
	 * @param string[] $paths
	 * @return \SplFileInfo[]|\Generator
	 */
	private function getFiles(array $paths): \Generator
	{
		foreach ($paths as $path) {
			if (\is_file($path)) {
				yield $path;

			} else {
				yield from Finder::findFiles($this->accepted)
					->exclude($this->excluded)
					->from($path)
					->exclude($this->excluded);
			}
		}
	}


	private function checkFile(File $file): bool
	{
		\set_error_handler(static function ($severity, $message, $file, $line): void {
			throw new \ErrorException($message, 0, $severity, $file, $line);
		});

		try {
			foreach ($this->checks as $check) {
				$check->check($file);
			}

			$file->finishedCheck();

		} catch (\Throwable $error) {
			$file->setCheckError($error);
		}

		\restore_error_handler();
		return $file->isOk();
	}

}
