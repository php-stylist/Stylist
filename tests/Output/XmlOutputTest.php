<?php declare(strict_types = 1);

namespace Stylist\Tests\Output;

use Stylist\CheckResult;
use Stylist\File;
use Stylist\Output\XmlOutput;
use Stylist\Tests\DummyCheck;
use Stylist\Tokenista\Tokens;
use Tester\Assert;
use Tester\FileMock;
use Tester\TestCase;


require_once __DIR__ . '/../bootstrap.php';


/**
 * @testCase
 */
final class XmlOutputTest extends TestCase
{

	public function testOutput(): void
	{
		$xml = FileMock::create('');

		$xmlOutput = new XmlOutput($xml);
		$xmlOutput->initialize(['foo']);

		$files = [];
		$tokens = Tokens::from('<?php');

		$files[] = $file1 = new File(new \SplFileInfo('foo/a'), $tokens, []);
		$xmlOutput->checkedFile($file1);

		$files[] = $file2 = new File(new \SplFileInfo('foo/b'), $tokens, []);
		$file2->addIssue(new DummyCheck(), 'Error', 5);
		$xmlOutput->checkedFile($file2);

		$files[] = $file3 = new File(new \SplFileInfo('foo/c'), $tokens, []);
		$file3->addIssue(new DummyCheck(), 'Error 1', 4);
		$file3->addIssue(new DummyCheck(), 'Error 2', 12);
		$xmlOutput->checkedFile($file3);

		$files[] = $file4 = new File(new \SplFileInfo('foo/d'), $tokens, []);
		$file4->addIssue(new DummyCheck(), 'Error 1', 14);
		$file4->addIssue(new DummyCheck(), 'Error 2', 29);
		$file4->addIssue(new DummyCheck(), 'Error 3', 3);
		$xmlOutput->checkedFile($file4);

		$files[] = $file5 = new File(new \SplFileInfo('foo/e'), $tokens, []);
		$file5->setCheckError(new \LogicException('Serious error during file check!'));
		$xmlOutput->checkedFile($file5);

		$files[] = $file6 = new File(new \SplFileInfo('foo/f'), $tokens, []);
		$xmlOutput->checkedFile($file6);

		$result = new CheckResult(false, $files, 0.5);
		$xmlOutput->finish($result);

		Assert::same(
			\file_get_contents(__DIR__ . '/XmlOutputTest.expected.xml'),
			\file_get_contents($xml)
		);
	}

}


(new XmlOutputTest())->run();
