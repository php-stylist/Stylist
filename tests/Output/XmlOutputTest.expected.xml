<?xml version="1.0"?>
<phpcs version="1.0.0">
  <file name="foo/a" errors="0" warnings="0"/>
  <file name="foo/b" errors="1" warnings="0">
    <error line="5" source="Stylist\Tests\DummyCheck" severity="5">Error</error>
  </file>
  <file name="foo/c" errors="2" warnings="0">
    <error line="4" source="Stylist\Tests\DummyCheck" severity="5">Error 1</error>
    <error line="12" source="Stylist\Tests\DummyCheck" severity="5">Error 2</error>
  </file>
  <file name="foo/d" errors="3" warnings="0">
    <error line="3" source="Stylist\Tests\DummyCheck" severity="5">Error 3</error>
    <error line="14" source="Stylist\Tests\DummyCheck" severity="5">Error 1</error>
    <error line="29" source="Stylist\Tests\DummyCheck" severity="5">Error 2</error>
  </file>
  <file name="foo/e" errors="1" warnings="0">
    <error line="0" source="LogicException" severity="5">Could not check further due to uncaught LogicException: Serious error during file check!</error>
  </file>
  <file name="foo/f" errors="0" warnings="0"/>
</phpcs>
